File: mcp/test/fixtures/bugfix-py/hits/real_bugs.py
"""The REAL-BUGS CORPUS, written by the AUDITOR rather than by the rules'
author, and the structural answer to how a wave of false-positive work can
open a false-negative hole with a green suite.

Everything else in hits/ is one minimal instantiation per rule, written by
whoever wrote the rule. That proves a rule fires at all. It cannot prove that
an exclusion added later did not eat a real bug, because a minimal hit fixture
carries no guard shapes for an exclusion to catch on — and the near-miss
fixtures only ever measure the direction the exclusion was written for.

Every defect below was SILENT before this round, and each one sits next to the
guard shape its rule's exclusions match, so widening any of those exclusions
by one step turns this file red.
"""
import asyncio
import os
import pathlib
import re


# --- race-condition-toctou-exists-open: four spellings that dominate modern
# --- code and that the two-exact-function-names version could not see.


def toctou_isfile(path):
    if os.path.isfile(path):
        return open(path).read()
    return ""


def toctou_access(path):
    if os.access(path, os.R_OK):
        return open(path).read()
    return ""


def toctou_negated_guard(path):
    if not os.path.exists(path):
        return ""
    return open(path).read()


def toctou_pathlib(path):
    p = pathlib.Path(path)
    if p.exists():
        return p.open().read()
    return ""


# --- null-safety-none-deref-dict-get: two real dictionary bugs that the old
# --- receiver allow-list silenced by SUBSTRING. A Flask/Django session IS a
# --- dict, and `client_config` is not an HTTP client.


def session_user_id(session):
    return session.get("user_id").strip()


def client_timeout(client_config):
    return client_config.get("timeout").total_seconds()


# --- null-safety-none-deref-match: three accessors that blow up in exactly
# --- the same way as `.group()` and were all silent.


def match_groups(text):
    return re.match(r"v(\d+)", text).groups()


def match_subscript(text):
    return re.match(r"v(\d+)", text)[1]


def match_span_start(text):
    return re.search(r"\d+", text).start()


# --- off-by-one-range-len-plus-one: the explicit-start spelling. The arity of
# --- `range` is part of the pattern, so `range(0, n + 1)` was invisible.


def explicit_start(values):
    acc = 0
    for i in range(0, len(values) + 1):
        acc += values[i]
    return acc


# --- edge-case-queryset-n-plus-one: a real N+1 behind `.exclude()`, and a
# --- real N+1 behind `.only()`. Both were silent, and for the wrong reason:
# --- ANY chained call broke the `$M.objects.all()` anchor, so the pack's
# --- `select_related` near-misses would have stayed green against a
# --- deliberately broken rule.


def n_plus_one_behind_exclude():
    out = []
    for book in Book.objects.exclude(draft=True):
        out.append(book.author.name)
    return out


def n_plus_one_behind_only():
    out = []
    for book in Book.objects.all().only("title"):
        out.append(book.author.name)
    return out


# --- error-handling-get-without-doesnotexist: an UNGUARDED get inside the
# --- `except` arm. The guard protects the `try` arm and nothing else, but a
# --- pattern-not-inside suppresses the whole try node, so this was silenced
# --- by the very guard that does not cover it.


def unguarded_in_except(user_id, fallback_id):
    try:
        return User.objects.get(pk=user_id)
    except User.DoesNotExist:
        return Profile.objects.get(pk=fallback_id)


# --- error-handling-except-pass and -bare-except: adding a `finally:`, an
# --- `else:` or a second exception type to the SAME swallowing `try`
# --- silenced both rules completely. Four swallows, unchanged in substance.


def pass_with_tuple(conn):
    try:
        conn.commit()
    except (ValueError, TypeError):
        pass


def pass_with_finally(conn):
    try:
        conn.commit()
    except ValueError:
        pass
    finally:
        conn.close()


def bare_with_finally(path):
    try:
        return read(path)
    except:
        return None
    finally:
        cleanup()


def bare_with_else(path):
    try:
        v = read(path)
    except:
        return None
    else:
        return v


# --- error-handling-get-without-doesnotexist, the other direction: the try
# --- exists but the handler does NOT guard the miss. Each of the three
# --- exclusion clauses carries a metavariable-regex that filters the caught
# --- TYPE, and without one of these three the filter is dead — the exclusion
# --- would be "any try/except at all silences the rule".


def get_in_try_wrong_handler(user_id):
    try:
        return User.objects.get(pk=user_id)
    except ValueError:
        return None


def get_in_try_wrong_handler_as(user_id):
    try:
        return User.objects.get(pk=user_id)
    except ValueError as exc:
        log(exc)
        return None


def get_in_try_wrong_handler_else(user_id):
    try:
        u = User.objects.get(pk=user_id)
    except ValueError:
        return None
    else:
        return u.name


# --- error-handling-except-pass: one function per remaining handler-shape x
# --- try-shape combination. Ablation found five of the nine branches dead
# --- without them, and a branch nothing exercises can be deleted while the
# --- suite stays green — which is how "add `else:` and the rule goes quiet"
# --- gets reintroduced.


def pass_with_else(conn):
    try:
        conn.commit()
    except ValueError:
        pass
    else:
        conn.log()


def pass_as_with_finally(conn):
    try:
        conn.commit()
    except ValueError as exc:
        pass
    finally:
        conn.close()


def pass_as_with_else(conn):
    try:
        conn.commit()
    except ValueError as exc:
        pass
    else:
        conn.log()


def bare_pass_with_finally(conn):
    try:
        conn.commit()
    except:
        pass
    finally:
        conn.close()


def bare_pass_with_else(conn):
    try:
        conn.commit()
    except:
        pass
    else:
        conn.log()


# --- null-safety-none-deref-match: the three remaining accessors. Ablation
# --- found their branches dead without a fixture each.


def match_groupdict(text):
    return re.search(r"(?P<n>\d+)", text).groupdict()


def match_end(text):
    return re.search(r"\d+", text).end()


def match_span(text):
    return re.search(r"\d+", text).span()


# --- race-condition-toctou-exists-open: the negated `isfile` guard. One
# --- branch per guard spelling, and this one had no fixture behind it.


def toctou_negated_isfile(path):
    if not os.path.isfile(path):
        return ""
    return open(path).read()


# --- memory-leak-open-without-context: a leaked handle in the same function
# --- as one whose ownership is transferred. The rule carries eight exclusion
# --- clauses and had no corpus entry at all; every one of those clauses
# --- unifies on the handle variable, and widening any of them to "some
# --- handle in this function escapes" would swallow this.


def leak_beside_transfer(path, rows):
    kept = open(path)
    leaked = open(path + ".log", "w")
    for row in rows:
        leaked.write(row)
    return kept


# --- race-condition-asyncio-not-awaited: an un-awaited coroutine in the same
# --- function as an awaited one, for the same reason.


async def unawaited_beside_awaited(task):
    await asyncio.sleep(1)
    asyncio.wait_for(task, timeout=5)
    return "done"

File: mcp/test/fixtures/bugfix-go/hits/real_bugs.go
// Package hits: the REAL-BUGS CORPUS, written by the AUDITOR rather than by
// the rules' author, and the structural answer to how a wave of
// false-positive work can open a false-negative hole with a green suite.
//
// Everything else in hits/ is one minimal instantiation per rule, written by
// whoever wrote the rule. That proves a rule fires at all. It cannot prove
// that an exclusion added later did not eat a real bug, because a minimal hit
// fixture carries no guard shapes for an exclusion to catch on — and the
// near-miss fixtures only ever measure the direction the exclusion was
// written for.
//
// Every defect below is deliberately placed NEXT TO the guard shape its
// rule's exclusions match, so that widening any of those exclusions by one
// step turns this file red:
//
//   - a leaked response in the same function as a closed one
//   - a discarded error in the same function as a `sync.Map.Load`
//   - an unguarded assertion on a DIFFERENT variable inside a type switch
//   - a ticker leaked while a sibling ticker is stopped
//   - an early return past an RUnlock, which the shipped rule could not see
//     at all
//   - the five HTTP entry points that dominate real clients and that the
//     `http.Get`-only anchor missed entirely
//   - a discarded error beside the Close idiom and a builtin discard
//   - a genuine off-by-one beside a correct n+1 DP seed
//
// Every rule in the pack has at least one entry here. The Java pack stated
// three rules at zero as a decision; this one has none at zero, because
// after this round every Go rule carries guard exclusions and none of them
// is low-risk by construction any more.
package hits

import (
	"io"
	"net/http"
	"os"
	"sync"
	"time"
)

type respBox struct {
	Headers map[string]string
}

// bug 1: the dominant real-world spelling. `client.Do(req)` is what anything
// setting headers, timeouts or contexts uses, and the shipped rule was
// anchored to `http.Get` alone.
func leakViaClientDo(c *http.Client, req *http.Request) error {
	resp, err := c.Do(req)
	if err != nil {
		return err
	}
	_ = resp.StatusCode
	return nil
}

// bug 2.
func leakViaPost(url string, body io.Reader) error {
	resp, err := http.Post(url, "application/json", body)
	if err != nil {
		return err
	}
	_ = resp.StatusCode
	return nil
}

// bug 3.
func leakViaDefaultClient(url string) error {
	resp, err := http.DefaultClient.Get(url)
	if err != nil {
		return err
	}
	_ = resp.StatusCode
	return nil
}

// bugs 3b and 3c: the remaining `http` package entry points. Ablation found
// both branches dead without a fixture behind them, and a branch nothing
// exercises can be deleted without a test moving — which is exactly how an
// anchor loses a spelling.
func leakViaPostForm(url string) error {
	resp, err := http.PostForm(url, nil)
	if err != nil {
		return err
	}
	_ = resp.StatusCode
	return nil
}

func leakViaHead(url string) error {
	resp, err := http.Head(url)
	if err != nil {
		return err
	}
	_ = resp.StatusCode
	return nil
}

// bug 4: a leak sitting beside a correct close in the SAME function. The
// close exclusions unify on the response variable; widen any of them to
// "there is a close somewhere" and this stops firing.
func leakBesideClose(url string, other string) error {
	good, err := http.Get(url)
	if err != nil {
		return err
	}
	defer good.Body.Close()
	bad, err := http.Get(other)
	if err != nil {
		return err
	}
	_ = bad.StatusCode
	return nil
}

// bug 5: a discarded error in the same function as a `sync.Map.Load`, whose
// second value really is a bool. The deny-list is per-callee; make it a
// per-function "this file talks to sync.Map" heuristic and this stops firing.
func discardBesideMapLoad(m *sync.Map, path string) []byte {
	_, _ = m.Load(path)
	data, _ := os.ReadFile(path)
	return data
}

// bug 6: an unguarded assertion on `w`, inside a type switch that proves
// nothing about `w` because it switches on `v`.
func assertOtherVarInTypeSwitch(v any, w any) string {
	switch v.(type) {
	case string:
		return w.(string)
	}
	return ""
}

// bug 7: an empty error branch whose variable is not spelled exactly `err`.
// The name filter must stay a substring test.
func emptyBranchNamedReadErr(path string) {
	readErr := os.Remove(path)
	if readErr != nil {
	}
}

// bug 8: a leaked ticker beside a stopped one. The stop exclusions unify on
// the ticker variable.
func tickerLeakBesideStop() {
	stopped := time.NewTicker(time.Second)
	defer stopped.Stop()
	leaked := time.NewTicker(time.Minute)
	<-leaked.C
}

// bug 9: an early return past `RUnlock`. `sync.RWMutex` read locks leak in
// exactly the same way and the shipped rule had no branch for them.
func rlockEarlyReturn(mu *sync.RWMutex, m map[string]int, key string) int {
	mu.RLock()
	if key == "" {
		return 0
	}
	v := m[key]
	mu.RUnlock()
	return v
}

// bug 10: the classic nil-map panic, which the `var $M map[$K]$V` anchor
// could not see: `&T{}` leaves every map field nil.
func structFieldNilMapWrite() *respBox {
	b := &respBox{}
	b.Headers["content-type"] = "application/json"
	return b
}

// bug 11: a discarded error beside the two shapes `err-blank-assign` now
// excludes — the Close idiom and a non-error builtin. The rule had NO corpus
// entry while carrying five exclusion clauses, which is exactly the profile
// the corpus exists for.
func discardBesideExcludedForms(f *os.File, path string) {
	_ = f.Close()                        // excluded: the Close idiom
	_ = copy(make([]byte, 1), []byte{2}) // excluded: builtin, no error
	_ = os.Remove(path)                  // fires: a real discarded error
}

// bug 12: a genuine off-by-one in the same function as a correct n+1 loop.
// The body requirement added in this round is what separates them, and
// without this the rule had no corpus entry at all.
func offByOneBesideDPSeed(xs []int) int {
	dp := make([]int, len(xs)+1)
	for i := 0; i <= len(xs); i++ {
		dp[i] = i // correct: dp has len(xs)+1 slots
	}
	sum := 0
	for i := 0; i <= len(xs); i++ {
		sum += xs[i] // BUG: indexes xs at len(xs)
	}
	return sum + dp[0]
}
